File: 2015/02/python/02.py
#!/usr/bin/env python
from operator import mul
from functools import reduce

#parse and sort box sides
def parse(line):
   line = line.strip().split('x')
   line = list(map(lambda x: int(x), line))
   return sorted(line)

#use the fact that sides are sorted
def paper(sides):
   paper = (
            3*sides[0]*sides[1] +
            2*sides[0]*sides[2] +
            2*sides[1]*sides[2]
            )
   return paper

def ribbon(sides):
   ribbon = (
             2*(sides[0]+sides[1]) +
             reduce(mul, sides, 1)
            )

   return ribbon

if __name__ == "__main__":
   with open('../input.txt', 'r') as f:
      input_list = f.readlines()
      input_list = [parse(box) for box in input_list]

   p1 = sum(map(paper , input_list))
   p2 = sum(map(ribbon, input_list))

   print(f"Part 1 answer: {p1}")
   print(f"Part 2 answer: {p2}")

File: 2015/02/ocaml/02.ml
#use "topfind"
#thread
#require "core.top"
#require "core.syntax"

open Core

(* parse each box dimensions into a sorted list *)

let parse_lines (lines: string list):  int list list = 
  let parse line = 
    String.split_on_chars ~on:[ 'x' ] line
    |> List.map ~f:int_of_string
    |> List.sort ~compare:Int.compare
  in
  List.map ~f:parse lines

(* calculation for part one *)

let paper_needed dims = 
  match dims with
  | s::m::l::_ -> 3*s*m + 2*s*l + 2*m*l
  | _       -> assert false

(* calculation for part two *)

let ribbon_needed dims = 
  match dims with
  | s::m::l::_ -> 2*(s+m) + (List.fold_left ~f:( * ) ~init:1 dims)
  | _       -> assert false

let () =
  let input  = In_channel.read_lines "../input.txt" in
  let parsed = parse_lines input in
  let paper  = List.map ~f:paper_needed parsed in
  let ribon  = List.map ~f:ribbon_needed parsed in
  let p1_ans = List.fold_left ~f:(+) ~init:0 paper in
  let p2_ans = List.fold_left ~f:(+) ~init:0 ribon in
    printf "Part 1 answer: %d\n" p1_ans;
    printf "Part 2 answer: %d\n" p2_ans;
